internal/backends/winit/muda.rs

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+use crate::SlintUserEvent;
+use i_slint_core::window::MenuVTable;
+use winit::event_loop::EventLoopProxy;
+use winit::window::Window;
+
+pub struct MudaAdapter {
+    entires: Vec<i_slint_core::items::MenuEntry>,
+    menubar: vtable::VBox<MenuVTable>,
+    // We need to keep menu alive, otherwise muda segfaults
+    _menu: muda::Menu,
+}
+
+impl MudaAdapter {
+    pub fn setup(
+        menubar: vtable::VBox<MenuVTable>,
+        winit_window: &Window,
+        proxy: EventLoopProxy<SlintUserEvent>,
+    ) -> Self {
+        type MenuRef<'a> = vtable::VRef<'a, MenuVTable>;
+        type EntryMap = Vec<i_slint_core::items::MenuEntry>;
+        fn generate_menu_entry(
+            menu: MenuRef,
+            entry: &i_slint_core::items::MenuEntry,
+            depth: usize,
+            map: &mut EntryMap,
+            window_id: &str,
+        ) -> Box<dyn muda::IsMenuItem> {
+            let id = muda::MenuId(format!("{window_id}|{}", map.len()));
+            map.push(entry.clone());
+            // the top level always has a sub menu regardless of entry.has_sub_menu
+            if !entry.has_sub_menu && depth != 0 {
+                Box::new(muda::MenuItem::with_id(
+                    id.clone(),
+                    &entry.title,
+                    true, /*entry.enabled*/
+                    None,
+                ))
+            } else {
+                let sub_menu =
+                    muda::Submenu::with_id(id.clone(), &entry.title, true /*entry.enabled*/);
+                if depth < 15 {
+                    let sub_entries = menu.sub_menu(Some(entry));
+                    for e in sub_entries {
+                        sub_menu
+                            .append(&*generate_menu_entry(menu, &e, depth + 1, map, window_id))
+                            .unwrap();
+                    }
+                } else {
+                    // infinite menu depth is possible, but we want to limit the amount of item passed to muda
+                    sub_menu
+                        .append(&muda::MenuItem::new(
+                            "<Error: Menu Depth limit reached>",
+                            false,
+                            None,
+                        ))
+                        .unwrap();
+                }
+                Box::new(sub_menu)
+            }
+        }
+
+        let menu = muda::Menu::new();
+        let mut map = EntryMap::new();
+        let menu_entries = menubar.sub_menu(None);
+        let window_id = u64::from(winit_window.id()).to_string();
+        for e in menu_entries {
+            menu.append(&*generate_menu_entry(menubar.borrow(), &e, 0, &mut map, &window_id))
+                .unwrap();
+        }
+
+        muda::MenuEvent::set_event_handler(Some(move |e| {
+            let _ = proxy.send_event(SlintUserEvent(crate::event_loop::CustomEvent::Muda(e)));
+        }));
```

Comment:
This will be called once per window, but I guess it should be called only once for the winit backend, so not quite in this place?

Reply:
Right. It could be called only once. But it is also not an issue to call it several times as it just replaces with itself.